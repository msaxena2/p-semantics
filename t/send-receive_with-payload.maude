in "../p-semantics.maude"

mod P-SEND-RECIEVE-TEST is including P-SEMANTICS .

  ops Sender Receiver Init SendRequests Done ePing receiver sent received message payload : -> Id .

  ops Receiver requestType : -> Type .
  ops SenderPgm ReceiverPgm : -> Stmt .

  eq SenderPgm = (
    machine Sender sends ePing ; {
      var receiver : Receiver ;
      var sent : int ;

      start state Init {
        entry {
          receiver = new Receiver( .Exps );
          goto SendRequests ;
        }
      }

      state SendRequests {
        entry {
          sent = sent + 1 ;
          send receiver, ePing, ( message = sent ) ;
        }
      }
    } ) .

   eq ReceiverPgm = (
    machine Receiver ; {
      var received : int ;
      start state Init {
        on ePing do (payload : requestType)  {
          received = payload . message ;
          goto Done ;
        }
      }
    } ) .

endm

red SenderPgm .
red ReceiverPgm .

red init(SenderPgm ReceiverPgm) .
rew init(main SenderPgm ReceiverPgm) .
--- search init(main SenderPgm ReceiverPgm) =>* C:Attributes .

quit .
