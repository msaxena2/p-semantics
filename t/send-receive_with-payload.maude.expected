==========================================
reduce in P-SEND-RECIEVE-TEST : SenderPgm .
result MachineStmt: machine Sender ; {
  var receiver : Receiver ;
  var sent : int ;
  start state Init {
    entry(.Exps){
      receiver = new Receiver(.Exps);
      goto SendRequests ;
    }
  }
  state SendRequests {
    entry(.Exps){
      sent = sent + 1 ;
      send receiver,ePing,message = sent ;
    }
  }
}
==========================================
reduce in P-SEND-RECIEVE-TEST : ReceiverPgm .
result MachineStmt: machine Receiver ; {
  var received : int ;
  start state Init {
    on ePing do(payload : requestType){
      received = payload . message ;
      goto Done ;
    }
  }
}
==========================================
reduce in P-SEND-RECIEVE-TEST : init(SenderPgm
ReceiverPgm) .
result Configuration: 
(pgm:  machine Sender ; {
    var receiver : Receiver ;
    var sent : int ;
    start state Init {
      entry(.Exps){
        receiver = new Receiver(.Exps);
        goto SendRequests ;
      }
    }
    state SendRequests {
      entry(.Exps){
        sent = sent + 1 ;
        send receiver,ePing,message = sent ;
      }
    }
  }
  machine Receiver ; {
    var received : int ;
    start state Init {
      on ePing do(payload : requestType){
        received = payload . message ;
        goto Done ;
      }
    }
  }) 
(machines:
  .MachineSchemas)
==========================================
rewrite in P-SEND-RECIEVE-TEST : init(main SenderPgm
ReceiverPgm) .
result Configuration: 
(machines:
  (machine: Sender
    ,(start: Init)
    ,(states:
      (state: Init
        ,(actions:
          .Actions
        )
,(entry: .Exps,{
          receiver = new Receiver(.Exps);
          goto SendRequests ;
        }))
      
      (state: SendRequests
        ,(actions:
          .Actions
        )
,(entry: .Exps,{
          sent = sent + 1 ;
          send receiver,ePing,message = sent ;
        }))
      
    )

    ,(initState:
      
        receiver |-> 0       
        sent |-> 0      
    ))
   ;; (machine: Receiver
    ,(start: Init)
    ,(states:
      (state: Init
        ,(actions:
          (action: ePing,payload : requestType,{
            received = payload . message ;
            goto Done ;
          })
        )
)
      
    )

    ,(initState:
      
        received |-> 0      
    ))
  ) (counter: 2) 
(    
instances: (instance: 0,(code: .Stmt),(variables: 
      this |-> 0     
      receiver |-> 1     
      sent |-> 1    ),(state: SendRequests),(mid: Sender),(buffer: .Messages)) (instance: 1,(variables: 
      this |-> 1     
      received |-> 1     
      payload |-> message = 1    ),(state: Init),(mid: Receiver),(buffer: .Messages),enterState(Done))
  )
Bye.
